staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go

```diff
@@ -927,7 +927,7 @@ func TestCacherDontMissEventsOnReinitialization(t *testing.T) {
 				for i := 12; i < 18; i++ {
 					w.Add(makePod(i))
 				}
-				w.Stop()
+				// Keep the watch open to avoid another reinitialization.
```

Comment:
maybe put in a `t.Cleanup(func() { w.Stop() })` just to make sure we don't have a leaked channel / goroutine?

Reply:
I was poking at this as well and ended up in a similar place for the fix, but noticed a few things the test was allowing which seemed questionable

1. if unexpected list / watch calls are made, the test doesn't notice
2. if all watchers encounter errors starting or watching and never receive a single watch event, the test doesn't notice

This is what I ended up with to ensure no unexpected list / watch calls are made, and at least a single rv watch event is received. I also got rid of the error channel and just errored the test directly to simplify things.

```diff
diff --git a/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go b/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go
index 7711bbcdf07..896b17aabfc 100644
--- a/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go
+++ b/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go
@@ -26,6 +26,7 @@ import (
 	"strconv"
 	"strings"
 	"sync"
+	"sync/atomic"
 	"testing"
 	"time"
 
@@ -901,10 +902,13 @@ func TestCacherDontMissEventsOnReinitialization(t *testing.T) {
 			var err error
 			switch listCalls {
 			case 0:
+				t.Log("list 0")
 				podList.ListMeta = metav1.ListMeta{ResourceVersion: "1"}
 			case 1:
+				t.Log("list 1")
 				podList.ListMeta = metav1.ListMeta{ResourceVersion: "10"}
 			default:
+				t.Error("unexpected list call", listCalls)
 				err = fmt.Errorf("unexpected list call")
 			}
 			listCalls++
@@ -915,6 +919,7 @@ func TestCacherDontMissEventsOnReinitialization(t *testing.T) {
 			var err error
 			switch watchCalls {
 			case 0:
+				t.Log("watch 0")
 				w = watch.NewFakeWithChanSize(10, false)
 				for i := 2; i < 8; i++ {
 					w.Add(makePod(i))
@@ -923,12 +928,15 @@ func TestCacherDontMissEventsOnReinitialization(t *testing.T) {
 				w.Error(nil)
 				w.Stop()
 			case 1:
+				t.Log("watch 1")
 				w = watch.NewFakeWithChanSize(10, false)
 				for i := 12; i < 18; i++ {
 					w.Add(makePod(i))
 				}
-				w.Stop()
+				// Clean up once the test is done
+				t.Cleanup(func() { w.Stop() })
 			default:
+				t.Error("unexpected watch call", watchCalls)
 				err = fmt.Errorf("unexpected watch call")
 			}
 			watchCalls++
@@ -941,30 +949,41 @@ func TestCacherDontMissEventsOnReinitialization(t *testing.T) {
 	}
 	defer cacher.Stop()
 
-	concurrency := 1000
+	minConcurrency := 1000
 	wg := sync.WaitGroup{}
-	wg.Add(concurrency)
 
 	// Ensure that test doesn't deadlock if cacher already processed everything
 	// and get back into Pending state before some watches get called.
 	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
 	defer cancel()
 
-	errCh := make(chan error, concurrency)
-	for i := 0; i < concurrency; i++ {
+	received := int64(0)
+	for i := 0; i < minConcurrency || atomic.LoadInt64(&received) == 0; i++ {
+		wg.Add(1)
 		go func() {
 			defer wg.Done()
+			if t.Failed() {
+				// bail out early if the test already failed
+				return
+			}
 			w, err := cacher.Watch(ctx, "pods", storage.ListOptions{ResourceVersion: "1", Predicate: storage.Everything})
 			if err != nil {
 				// Watch failed to initialize (this most probably means that cacher
 				// already moved back to Pending state before watch initialized.
 				// Ignore this case.
+				if i > minConcurrency {
+					t.Logf("error on watch init (will retry): %v", err)
+				}
 				return
 			}
 			defer w.Stop()
 
 			prevRV := -1
 			for event := range w.ResultChan() {
+				if t.Failed() {
+					// bail out early if the test already failed
+					return
+				}
 				if event.Type == watch.Error {
 					break
 				}
@@ -974,24 +993,20 @@ func TestCacherDontMissEventsOnReinitialization(t *testing.T) {
 				}
 				rv, err := strconv.Atoi(object.(*example.Pod).ResourceVersion)
 				if err != nil {
-					errCh <- fmt.Errorf("incorrect resource version: %v", err)
+					t.Errorf("incorrect resource version: %v", err)
 					return
 				}
 				if prevRV != -1 && prevRV+1 != rv {
-					errCh <- fmt.Errorf("unexpected event received, prevRV=%d, rv=%d", prevRV, rv)
+					t.Errorf("unexpected event received, prevRV=%d, rv=%d", prevRV, rv)
 					return
 				}
 				prevRV = rv
+				atomic.AddInt64(&received, 1)
 			}
-
 		}()
 	}
 	wg.Wait()
-	close(errCh)
-
-	for err := range errCh {
-		t.Error(err)
-	}
+	t.Log("received", received)
 }
 
 func TestCacherNoLeakWithMultipleWatchers(t *testing.T) {

```

Reply:
> if all watchers encounter errors starting or watching and never receive a single watch event, the test doesn't notice

I'm aware of it, and indeed it may sometimes happen, but that's rare
the nature of the test is somewhat probabilities anyway, so I would say that's ok

Reply:
Added a cleanup func as suggested and updated to call t.Errorf instead of having channel of errors.